proposals/threads/Globals.md

```diff
@@ -2,6 +2,341 @@
+* The SP is actually in linear memory, so it can easily be trashed by another
+  agent
+
+### Solution 2: Use Internal Mutable Global w/ Shadow in Linear Memory
+
+To reduce the cost of accessing SP, we can store the SP in a mutable global. To
+make this work across module boundaries, we spill the SP to a linear memory in
+the caller and load the SP in the callee.
+
+This could be optimized to only spill SP when necessary (although this is
+complicated by indirect function calls), but for simplicity this example will
+just show spilling before function calls and loading at function entries:
+
+```
+(module $m1
+  (import "env" "shadowSpAddr" (global $shadowSpAddr i32))
+  (memory (export "memory") 1)
+  (global $sp (mut i32) (i32.const 0))
+
+  (data (i32.const 0) "\00\01\00\00")  ;; Shadow SP for agent 0
+  (data (i32.const 4) "\00\02\00\00")  ;; Shadow SP for agent 1
+
+  (func
+    ;; Load shadow SP
+    (set_global $sp (i32.load (get_global $shadowSpAddr)))
+
+    ;; SP = SP + 64
+    (set_global $sp (i32.add (get_global $sp) (i32.const 64)))
+    ...
+    ;; Function call, spill SP
+    (i32.store (get_global $shadowSpAddr) (get_global $sp))
+  )
+  ...
+)
+
+(module $m2
+  (import "env" "memory" (memory 1))
+  (import "env" "spAddr" (global $shadowSpAddr i32))
+
+  (global $sp (mut i32) (i32.const 0))
+
+  (func
+    ;; Load shadow SP
+    (set_global $sp (i32.load (get_global $shadowSpAddr)))
+    ...
+  )
+  ...
+)
+```
+
+The modules would be instantiated the same as they would in solution 1 above:
+
+```
+let agentIndex = 0;
+let shadowSpAddrs = [0x0, 0x4];
+let imports = {env: {shadowSpAddr: shadowSpAddrs[agentIndex]}};
+WebAssembly.instantiate(m1Bytes, imports).then(
+    ({instance}) => {
+        let imports = {env: {
+          memory: instance.exports.memory,
+          shadowSpAddr: shadowSpAddrs[agentIndex],
+        }};
+        WebAssembly.instantiate(m2Bytes, imports).then(...);
+    });
+```
+
+This solution could be extended to other thread-local variables, but would
+have additional overhead for every function call. It's likely to only be
+valuable for thread-local values that are used often. In all other cases, it
+would be best to just use the shadow values directly.
+
+This solution has the following drawbacks:
+
+* Mostly just a (potential) optimization of solution 1
+* All function calls that load/store SP must spill SP at call sites and
+  function entrypoints
+* Additional thread-local values must also be spilled/loaded in the same way to
+  have the same benefit
+* The SP is still in linear memory, so it can easily be trashed by another
+  agent
+
+### Solution 3: Modify Function Signature to Pass SP as Parameter
+
+Rather than storing spilling the SP to linear memory, the SP value can be
+passed as a parameter. Because we can't tell whether an imported function will
+use the SP, we must modify all imported and exported functions.
+
+The SP will ultimately be saved in a mutable global, but will be loaded from
```

Comment:
It's just an optimization. You could always use the parameter, but the idea here was to only modify exported functions' signatures.